tt_eager/tt_dnn/op_library/composite/composite_ops.cpp

```diff
@@ -714,12 +724,27 @@ Tensor subalpha(const Tensor& input_a, const Tensor& input_b, float alpha, const
 }
 
 // addalpha(input, other, alpha) = input + (alpha * other)
-Tensor _addalpha(const Tensor& input_a, const Tensor& input_b, float alpha, const MemoryConfig& output_mem_config) {
-    Tensor result = add(mul_unary(input_b, alpha, output_mem_config), input_a, std::nullopt, output_mem_config);
-    return result;
+Tensor _addalpha(
+    const Tensor& input_a,
+    const Tensor& input_b,
+    float alpha,
+    std::optional<Tensor*> output_tensor,
+    const MemoryConfig& output_mem_config) {
+    if (output_tensor.has_value() && *output_tensor) {
+        *(*output_tensor) = add(mul_unary(input_b, alpha, output_mem_config), input_a, std::nullopt, output_mem_config);
```

Comment:
```suggestion
    std::optional<Tensor> output_tensor,
    const MemoryConfig& output_mem_config) {
    if (output_tensor.has_value()) {
        output_tensor = add(mul_unary(input_b, alpha, output_mem_config), input_a, output_tensor, output_mem_config);
```

The output_tensor object provided by the user should be used and remain active. Using *(*output_tensor) will remove the existing output_tensor. The output_tensor variable will be replaced by the newly created tensor object by the add function.

Reply:
@dongjin-na  passing output_tensor to the add op is what I would ideally like to do. 
`add(mul_unary(input_b, alpha, output_mem_config), input_a, std::nullopt, output_mem_config, output_tensor);`
Until #8394 is merged both composite and backward ops implementation will have to be put on hold.

Reply:
Okay. If that's what you have planned, I think it would be better to use std::optional<Tensor> and return a new Tensor object instead of using std::optional<Tensor*>. It might be better to stick with the existing approach.

What I mean is that we should keep using std::optional<Tensor> output_tensor as a parameter before #8394 is merged. Once #8394 is integrated, we can make the necessary adjustments, focusing only on the function's internal implementation without changing the interface. This approach seems more efficient to me.

Anyway, it seems like we are on the same page regarding the final task, so I'm fine with either implementation.

Reply:
Hi @dongjin-na and @eyonland , 

`t3 = ttl.tensor.addalpha(t0, t1, alpha, output_tensor=t2, output_mem_config=output_mem_config)`

Using `std::optional<Tensor> output_tensor` in composite structure passes the tensor by value and the result doesn't reflect in output_tensor(t2) after the op is executed. Even though output_tensor is used for computation, we can only return the result as a new Tensor object(t3). Are you okay with this approach ? 
![image (50)](https://github.com/tenstorrent/tt-metal/assets/156762498/51cbf032-481c-4645-8ee8-c1922ec8fe1a)

I tried using reference_wrapper and was able to do modify output_tensor in memory. But I remember something about not using std::reference_wrapper, (Eyon mentioned something) not sure why but here are the results though. Whichever approach you are ok with, we'll go with that.

Add optional output tensor to addalpha op using `std::optional<std::reference_wrapper<Tensor>> output_tensor`
- The passed output_tensor is modified inside the op with the expected result. It is used for computation without creating extra tensors.

<img width="1187" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/156762498/3e151921-f2cc-4e26-881b-b93dc83bcca8">

- Using binary op with output_tensor
<img width="1512" alt="Screenshot 2024-06-02 at 3 45 28 AM" src="https://github.com/tenstorrent/tt-metal/assets/156762498/8f39217f-66d4-40fc-b401-ba403a25c242">